commands/start.js

```diff
@@ -23,32 +26,55 @@ function main(urc) {
   logger.log(
     `Starting underreact in ${urc.mode} mode. ${chalk.yellow('Wait ...')}`
   );
-  return (
-    del(urc.outputDirectory, { force: true })
-      // webpack needs to run first as others depend on the assets
-      .then(() => new Promise(resolve => watchWebpack(urc, resolve)))
-      .then(() => {
-        Promise.all([
-          writeHtml(urc),
-          writeCss(urc),
-          autoCopy.copy({
-            sourceDir: urc.publicDirectory,
-            destDir: urc.outputDirectory
-          })
-        ]);
-      })
-      .then(() => startServer(urc))
-      .then(() => {
-        watchCss(urc);
-        watchPublicDir(urc);
-      })
+
+  return del(urc.outputDirectory, { force: true }).then(() =>
+    watch(urc, () => startServer(urc))
   );
 }
 
-function watchWebpack(urc, onFirstRun) {
+function watch(urc, onFirstWrite) {
+  let cssOutput;
+  let webpackAssets;
+  let isFirst = true;
+  const htmlCachedWrite = writeHtml(urc);
+  const hasStylesheets = urc.stylesheets.length > 0;
+  const htmlWriter = () => {
+    if (isFirst) {
+      onFirstWrite();
+      isFirst = false;
+    }
+    htmlCachedWrite(
+      new Assets({
+        urc,
+        cssOutput,
+        webpackAssets
+      })
+    );
+  };
+
+  watchPublicDir(urc);
+
+  if (hasStylesheets) {
+    watchCss(urc, output => {
+      cssOutput = output;
+      if (webpackAssets) {
```

Comment:
I fear this is confusing state management. Can you explain why `watchCss` and `watchWebpack` may or may not end up calling `htmlWriter`; and why `watchCss` and `watchWebpack` are so entangled?

Reply:
I tried really hard on this particular problem, but as you can see it is a bit of entangled mess.
So both `watchCss` and `watchWebpack` can independently start, but there are certain conditions that need to be met before we can write HTML.
1. watchCSS needs to return the final url of the output css.
2. Webpack needs to finish.
3. Webpack can go ahead and write html if there are no stylesheets.
4. Once html is written we need to signal server to start serving files.

You got any tips on how to untangle this ?

Reply:
🤔 A few thoughts:

- This and other complexities might be an argument for leaning more on Webpack — if that ends up actually reducing complexity? I dunno.
- This is all in order to include hashes during `start`, right? This complexity may be a reason *not* to do that.

I definitely do think that this is too confusing to keep as is — we should sort out something less tangled before settling.